### PR TITLE
Performance improvement for Array#intersection

### PR DESCRIPTION
**Pull request by travisofthenorth**
Performance improvement mentioned here: https://github.com/ruby/ruby/pull/2533

Instead of creating N intermediate arrays, this repurposes the current implementation for `&` as the new implementation for `intersection` and holds onto a single result array.

The algorithm is:

1. Find the shortest array, uniq it, use that as the running result set
2. Loop through the arrays, and perform the `&` implementation, i.e.
- if both arrays are relatively small, try to find each element of one array in the other, otherwise
- hash the array and check each element in the result set against the hash

### CODE REVIEW

File: array.c [hunk 1]
```diff
@@ -4636,6 +4636,10 @@ rb_ary_difference_multi(int argc, VALUE *argv, VALUE ary)
 static VALUE
 rb_ary_and(VALUE ary1, VALUE ary2)
 {
+    // VALUE arg[1];
+    // arg[0] = ary2;
+    // return rb_ary_intersection_multi(1, arg, ary1);
```
Comment (travisofthenorth) [inline array.c]: For some reason, when I try to rewrite `&` in terms of `intersection`, I'm getting a weird error:
```
Traceback (most recent call last):
	7: from ./tool/transform_mjit_header.rb:5:in `<main>'
	6: from ./tool/transform_mjit_header.rb:5:in `require'
	5: from /Users/travis/code/ruby/lib/fileutils.rb:106:in `<top (required)>'
	4: from /Users/travis/code/ruby/lib/fileutils.rb:1627:in `<module:FileUtils>'
	3: from /Users/travis/code/ruby/lib/fileutils.rb:1627:in `inject'
	2: from /Users/travis/code/ruby/lib/fileutils.rb:1627:in `each'
	1: from /Users/travis/code/ruby/lib/fileutils.rb:1628:in `block in <module:FileUtils>'
/Users/travis/code/ruby/lib/fileutils.rb:1628:in `instance_method': undefined method `private_module_function' for module `FileUtils' (NameError)
```

File: array.c [hunk 2]
```diff
@@ -5101,6 +5075,71 @@ rb_ary_compact(VALUE ary)
...
+rb_ary_intersection_multi(int argc, VALUE *argv, VALUE ary)
+{
+    VALUE hash, v, result, current, shortest;
+    st_data_t vv;
+    long i, j, min;
+
+    shortest = ary;
+    min = RARRAY_LEN(ary);
+    for (i = 0; i < argc; i++) {
+        current = to_ary(argv[i]);
```
Comment (travisofthenorth) [inline array.c]: @nobu perhaps you could advise me. I'd like to sort `argv` by the size of the input arrays, but I'm not sure how to accomplish that with something like `rb_ary_sort` which seems to sort them recursively.

Comment (nobu) [reply array.c]:
```c
static int
sort_by_size(const void *a, const void *b, void *dummy)
{
    long alen = RARRAY_LEN(*(const VALUE *)a);
    long blen = RARRAY_LEN(*(const VALUE *)b);
    if (alen < blen) return +1;
    if (alen > blen) return -1;
    return 0;
}

{
    // ...
    for (i = 0; i < argc; i++) argv[i] = to_ary(argv[i]);
    ruby_qsort(argv, argc, sizeof(VALUE), sort_by_size, NULL);
    // ...
}
```

Comment (travisofthenorth) [reply array.c]: Thank you!

**Pull request closed without merging**